README.md

```diff
@@ -0,0 +1,110 @@
+        fmt.Println("Starting server on :8080...")
+        if err := http.ListenAndServe(":8080", nil); err != nil {
+            log.Fatalf("Server error: %v", err)
+        }
+    }
+   ```
+
+2. The `handleWebSocket` performs the protocol upgrade and assigns various websocket handler functions, such as `OnOpen` or `OnMessage`:
+
+   ```go
+    func handleWebSocket(w http.ResponseWriter, r *http.Request) {
+        conn, err := upgrader.Upgrade(w, r, nil)
+        if err != nil {
+            log.Printf("Error upgrading to websocket: %v", err)
+            return
+        }
+        handlers.OnOpen(conn)
+
+        go func() {
+            defer handlers.OnClose(conn)
+            for {
+                messageType, message, err := conn.ReadMessage()
+                if err != nil {
+                    handlers.OnError(conn, err)
+                    break
+                }
+                handlers.OnMessage(conn, messageType, message)
+            }
+        }()
+    }
+   ```
+
+3. The WebSocket application logic is located in the `pkg/handlers/handlers.go` file and contains callbacks for each WebSocket event:
+
+   ```go
+    func OnOpen(conn *websocket.Conn) {
+        log.Printf("WebSocket connection opened: %v", conn.RemoteAddr())
+    }
+
+    func OnMessage(conn *websocket.Conn, messageType int, message []byte) {
+        log.Printf("Received message from %v: %s", conn.RemoteAddr(), string(message))
+
+        if err := conn.WriteMessage(messageType, message); err != nil {
+            log.Printf("Error sending message: %v", err)
+        }
+    }
+
+    func OnClose(conn *websocket.Conn) {
+        log.Printf("WebSocket connection closed: %v", conn.RemoteAddr())
+        conn.Close()
+    }
+
+    func OnError(conn *websocket.Conn, err error) {
+        log.Printf("WebSocket error from %v: %v", conn.RemoteAddr(), err)
+    }    
+   ```
+
+## Build and Deploy the application
+
+At the the root of directory run the following command:
+
+   ```bash
+   ko apply -f config/service.yaml
+   ```
+
+## Testing the WebSocket server
+
+Get the URL for your Service with:
+
+   ```bash
+   k get ksvc
+   NAME               URL                                                 LATESTCREATED            LATESTREADY              READY   REASON
+   websocket-server   http://websocket-server.default.svc.cluster.local   websocket-server-00001   websocket-server-00001   True    
+   ```
+
+Now run a container with the [wscat](https://github.com/websockets/wscat) CLI and point it to the WebSocket application `ws://websocket-server.default.svc.cluster.local/ws`, like:
+
+
+   ```bash
+   kubectl run --rm -i --tty wscat --image=monotykamary/wscat --restart=Never -- -c ws://websocket-server.default.svc.cluster.local/ws
+   ```
+
+Afterward you can chat with the WebSocket server like:
+
+   ```bash
+   ```If you don't see a command prompt, try pressing enter.
+   ```connected (press CTRL+C to quit)
+   ```> Hello
+   ```< Hello
+   ```> 
+   ```
```

Comment:
We can add a note on autoscaling and show some traffic and pods scaled. Websockets autoscaling is based on connections. Could we configure autoscaling.knative.dev/target to some number and show how pods are being autoscaled. This could also be done in some other PR as we dont have that info for grpc either.